### PR TITLE
MET-137: move constraint validation into respective directory

### PR DESCRIPTION
**Pull request by StevenVAIDIE**
**Description (for Contributor and Core Developer)**

- We splitted the Validators that were applied to the command classes in their respective directories : `Validation/CreateMeasurementFamily` and `Validation/SaveMeasurementFamily` instead of only `Validation/MeasurementFamily`.
- We added the method `MeasurementFamily::normalizeWithIndexedUnits` into the model.
- We updated the Save controller to switch between a `CreateMeasurementFamilyCommand` and `SaveMeasurementFamilyCommand` depending on the request.
- The Save controller supports partial updates
- We rewrited the tests for the External API Save endpoint with the following use cases:
  - it_create_a_measurement_family_when_it_does_not_exists
  - it_returns_an_error_when_the_measurement_family_does_not_have_a_code
  - it_returns_an_error_when_the_units_are_not_correctly_indexed
  - it_does_nothing_when_the_measurement_family_already_exists_and_is_exactly_the_same
  - it_add_an_unit_when_it_does_not_exist
  - it_add_a_label_translation
  - it_update_a_label_translation
  - it_update_an_unit_operations
  - it_creates_multiple_measurement_families
  - it_returns_an_error_when_the_measurement_family_list_does_not_have_the_right_structure
  - it_returns_an_error_when_the_measurement_family_creation_does_not_have_the_right_structure
  - it_returns_an_error_when_the_measurement_family_update_does_not_have_the_right_structure

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | -
| Added legacy Behats               | -
| Added acceptance tests            | -
| Added integration tests           | -
| Changelog updated                 | -
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Todo
| Migration script                  | -
| Tech Doc                          | -

### CODE REVIEW

File: src/Akeneo/Tool/Bundle/MeasureBundle/Controller/ExternalApi/SaveMeasurementFamiliesAction.php [hunk 1]
```diff
@@ -38,18 +48,30 @@ class SaveMeasurementFamiliesAction
...
+        MeasurementFamilyRepository $measurementFamilyRepository
     ) {
         $this->measurementFamilyListValidator = $measurementFamilyListValidator;
+        $this->measurementFamilyCommonStructureValidator = $measurementFamilyCommonStructureValidator;
         $this->measurementFamilyStructureValidator = $measurementFamilyStructureValidator;
         $this->validator = $validator;
         $this->violationNormalizer = $violationNormalizer;
         $this->saveMeasurementFamilyHandler = $saveMeasurementFamilyHandler;
+        $this->createMeasurementFamilyHandler = $createMeasurementFamilyHandler;
+        $this->measurementFamilyRepository = $measurementFamilyRepository;
```
Comment (tseho) [inline src/Akeneo/Tool/Bundle/MeasureBundle/Controller/ExternalApi/SaveMeasurementFamiliesAction.php]: it is ok if we inject the repository there ? We need to check if the measurement family code exists to switch between `create` and `save` commands, and also guess the appropriate return code.

Comment (ValentinMumble) [reply src/Akeneo/Tool/Bundle/MeasureBundle/Controller/ExternalApi/SaveMeasurementFamiliesAction.php]: I guess so but why not the interface?

Comment (tseho) [reply src/Akeneo/Tool/Bundle/MeasureBundle/Controller/ExternalApi/SaveMeasurementFamiliesAction.php]: You are right, this is a mistake.

File: src/Akeneo/Tool/Bundle/MeasureBundle/Model/MeasurementFamily.php [hunk 2]
```diff
@@ -62,6 +62,20 @@ function (Unit $unit) {
         ];
     }
 
+    public function normalizeWithIndexedUnits(): array
```
Comment (ValentinMumble) [inline src/Akeneo/Tool/Bundle/MeasureBundle/Model/MeasurementFamily.php]: We decided to not put it in the model in this PR https://github.com/akeneo/pim-community-dev/pull/11834/files#diff-c2b785254b6fefa7e19c523bdb106c65L60 but I guess if we do, we should update it there too

Comment (tseho) [reply src/Akeneo/Tool/Bundle/MeasureBundle/Model/MeasurementFamily.php]: I do agree we should also update for the GET.
Why didn't you put it in the model ?

Comment (ValentinMumble) [reply src/Akeneo/Tool/Bundle/MeasureBundle/Model/MeasurementFamily.php]: As it was only for the external API we assumed it was more an infra layer-related issue than a model one but I really don't mind either solution

**Comment by StevenVAIDIE** [comment]
Currently, the API allow scientific annotation for convert_from_standard[value] on internal and external controller (we only have validation on front-end for internal controller with digit regex). 

When we add unit with scientific annotation we save the scientific annotation inside database. But i think that we cannot save the family on interface.
When we update scientific annotation with the same scientific annotation there is no specific problem, we save the scientific annotation inside database even if the measurement family is used on a attribute.
But, when we update float convert_from_standard[value], with scientific annotation and the unit is already assign to attribute, we throw an error because we consider that the scientific annotation is not the same than previous float.

Currently we have the problem when we launch the [migration project](https://github.com/akeneo/YamlToMeasurement) with this file https://github.com/akeneo/MeasureBundle/blob/master/Resources/config/measure.yml
In the "Length" family the "MIL" unit have a convert_from_standard[value]=0.0000254. The project YamlToMeasurement cast this float to 2.54E-5.

**Comment by tseho** [comment]
@StevenVAIDIE good catch, can you create a ticket in the sprint ?

**Pull request merged**